src-tauri/src/chaodna.rs

```diff
@@ -0,0 +1,76 @@
+use rand::{Rng, distributions::{Distribution, Standard}};
+use crate::primer::Base;
+// use diff::Diff;
+
+
+pub trait RandError {
+    fn modify_base(&self, initial_seq: &Vec<Base>, prob: f64) -> Vec<Base>; 
+    fn delete_base(&self, initial_seq: &Vec<Base>, prob: f64) -> Vec<Base>;
+    fn full_chaos(&self, initial_seq: &Vec<Base>, mod_prob: f64, del_prob: f64) -> Vec<Base>;
+//    fn diff_chaos(&self, input: &Vec<Base>, del_prob: f64, mod_prob: f64) -> Vec<diff::Result<Base>>;
+}
+
+
+pub struct Chaos {
+}
+
+
+impl RandError for Chaos {
+    // Takes a base vector and probability as input, outputs a base vector with modified bases
+    fn modify_base(&self, initial_seq: &Vec<Base>, prob: f64) -> Vec<Base> {
+	let mut rng = rand::thread_rng();
+	initial_seq
+	    .iter()
+	    .map(|base| {
+		// RNG boolean; if true, keeps running until a different base from the original is selected
+		if rng.gen_bool(prob) { 
+                    loop {
+			let new_base = rng.gen::<Base>();
+			if new_base != *base {
+                            return new_base; 
+			}
+                    }
+		    // If RNG boolean is false, returns original (no change)
+		} else {
+                    *base 
+		}
+            })
+	    .collect()
+    }
+
+    fn delete_base(&self, initial_seq: &Vec<Base>, prob: f64) -> Vec<Base> {
+	let mut rng = rand::thread_rng();
+	initial_seq
+	    .iter()
+	    .filter_map(|base| {
+		// RNG boolean; if true, returns nothing
+		if rng.gen_bool(prob) {
+                    None
+		}
+		    // If RNG boolean is false, returns original (no change)
+		else {
+                    Some(*base)
+		}
+            })
+	    .collect()
+    }
+
+    fn full_chaos(&self, initial_seq: &Vec<Base>, mod_prob: f64, del_prob: f64) -> Vec<Base> {
+	let mod_seq = self.modify_base(initial_seq, mod_prob);
+	self.delete_base(&mod_seq, del_prob)
+    }
+
```

Comment:
For full_chaos, I was thinking to simulate edit errors, we define $P_{edit}$ as the probability that an edit error occurs, then, we have $P_{edit} = P_{mutation} + P_{delete}$. The code may look something like this:

TLDR: While we map over a sequence of bases, we first randomly sample, does an edit error occur at all? And if it does, we randomly sample the type of edit error.
 
1. there is a probability that an edit error occurs, if edit error doesn't occur, we continue to the next base, if an edit error does occur, then we now have two discrete types of edit errors that can occur.
2. we randomly sample the two discrete types of errors, and then if deletion error occurs, we return no base, if mutation occurs, then we randomly sample again, which mutation occurs (should this be only of three bases of all four?).

The way your `full_chaos` is set up, we first randomly mutate, and then randomly delete; I'm not sure if this is the same definition of $P_{edit}$, and we might "delete" a mutated base, which I don't think would occur in synthesis? 

Let's discuss on Discord and in our software design meeting!

Reply:
Sure! Just to be clear, assuming $P_{edit}$ occurs, what would the probability of deletion/mutation be set to? Would those be defined in input? Would the input then be vector, $P_{edit}$, $P_{del}$, $P_{mut}$?

As for the order of `full_chaos`, this was an oversight on my part; I initially had it call `modify_base` and then `insert_base` after; I later swapped all instances of `insert_base` with `delete_base` when I rewrote the function and forgot to change the order.

Reply:
Yeah, those $P_{mut}$, $P_{del}$ would be defined, and after your modify your implementation then I think it will make sense.

Reply:
Sorry, $P_{edit}$, $P_{del}$ and $P_{mut}$ would all be defined. Let me know if this is unclear

Reply:
![image](https://github.com/UBC-iGEM/dna-software/assets/55033656/006868b9-04ce-4fd2-884e-c5a7cf9e2add)

![image](https://github.com/UBC-iGEM/dna-software/assets/55033656/b981f8e4-43c5-46b2-8201-e22181645de5)

Via [Short Systematic Codes for Correcting Random Edit Errors in DNA Storage](https://www.notion.so/ubcigem/Short-Systematic-Codes-for-Correcting-Random-Edit-Errors-in-DNA-Storage-cfd45e2b2a504c8f83494fad6747a91a?pvs=4)

Reply:
Ah I see. So $P_{edit}$ defines the error probability, and then $P_{del/mut}$ define the skew between deletion and mutation. What happens if the latter two don’t sum to 1? Should the program continue or should it return an error? If it should continue, then there is a chance that $P_{edit}$ occurs but there is no change…

Alternatively, we could just have one parameter for $P_{del}$ and then implicitly calculate $P_{mut}$ as $1-P_{del}$. In GUI there could be a slider to adjust the relative proportion of these two.

Reply:
We can implicitly calculate $P_{mut}$ based on $P_{del}$ and show a slider in the GUI

Reply:
I have pushed up this implementation to the branch.